src/Microsoft.Health.Fhir.Liquid.Converter/Utilities/ComplexObjectFilterUtility.cs

```diff
@@ -0,0 +1,173 @@
+                            var localPath = new Queue<string>(path);
+                            if (ObjHasValueAtPath(obj, localPath, value))
+                            {
+                                return true;
+                            }
+                        }
+
+                        return false;
+                    }
+
+                    // If our array key references a specific index
+                    else
+                    {
+                        // Trim and parse out our numeric index
+                        var indexStr = key.TrimStart('[').TrimEnd(']');
+                        var index = int.Parse(indexStr);
+
+                        if (index >= 0 && inputArray.Count() > index)
+                        {
+                            // Recurse
+                            return ObjHasValueAtPath(inputArray[index], path, value);
+                        }
+                        else
+                        {
+                            // It's not possible for this array to have the element specified
+                            // so this is not a match
+                            return false;
+                        }
+                    }
+                }
+                else
+                {
+                    return false;
+                }
+            }
+
+            // Our key is referencing an object property
+            else
+            {
+                if (input is Dictionary<string, object> inputObject)
+                {
+                    if (inputObject.ContainsKey(key))
+                    {
+                        // Recurse
+                        return ObjHasValueAtPath(inputObject[key], path, value);
+                    }
+                }
+
+                return false;
+            }
+        }
+
+        // This function splits a string into an array of parts like
+        // `test.path[].to[5].value` -> ['test', 'path', '[]', 'to', '[5]', 'value']
+        public static Queue<string> SplitObjectPath(string path)
+        {
+            var parts = new Queue<string>();
+            var part = new StringBuilder();
+            var bracket = false;
+
+            for (int i = 0; i < path.Length; i++)
+            {
+                if (Delimeters.Contains(path[i]) == false)
+                {
+                    // Any char that isn't a delimeter gets saved
+                    part.Append(path[i]);
+                }
+
+                if (path[i] == '[')
+                {
+                    if (part.Length > 0)
+                    {
+                        // Perform a split
+                        parts.Enqueue(part.ToString());
+                        part.Clear();
+                    }
+
+                    // Brackets do not get stripped
+                    part.Append(path[i]);
+
+                    // Begin a bracket split
+                    bracket = true;
+                }
+
+                if (path[i] == ']')
```

Comment:
Should this if and the one below be an `else if`?  I'm assuming we don't need to check each section if a previous one matched - as i isn't going to change.